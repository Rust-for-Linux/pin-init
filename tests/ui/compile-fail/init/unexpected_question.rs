#![feature(allocator_api)]
#![allow(unused_attributes)]
#[path = "../../../../examples/mutex.rs"]
mod mutex;
use mutex::*;
use pinned_init::*;
use std::pin::Pin;

#[pin_data]
struct Foo {
    next: Option<Pin<Box<Foo>>>,
}

impl Foo {
    fn new(next: Option<impl PinInit<Self, Error>>) -> impl PinInit<Self, Error> {
        try_pin_init!(Self {
            next: next.map(|next| Box::try_pin_init(next)).transpose()?,
        }? Error)
    }
}

#[pin_data]
struct Bar {
    #[pin]
    foo: Foo,
}

impl Bar {
    fn new() -> impl PinInit<Self, Error> {
        try_pin_init!(Self {
            foo <- Foo::new(None)?,
        }? Error)
    }
}

fn main() {}
