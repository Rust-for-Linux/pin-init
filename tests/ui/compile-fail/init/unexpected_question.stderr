error[E0277]: the `?` operator can only be applied to values that implement `Try`
  --> tests/ui/compile-fail/init/unexpected_question.rs:31:20
   |
31 |             foo <- Foo::new(None)?,
   |                    ^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `impl pinned_init::PinInit<Foo, mutex::linked_list::error::Error>`
   |
   = help: the trait `Try` is not implemented for `impl pinned_init::PinInit<Foo, mutex::linked_list::error::Error>`
